src/Plugin/Components/SettingsContainer.ts

```diff
@@ -227,16 +218,159 @@ export class SettingsContainer {
 			.setName("Tokens")
 			.setDesc("The number of tokens used in the completion.")
 			.addText((text) => {
-				text.setValue(`${this.plugin.settings.tokens}`);
+				text.setValue(`${viewSettings.chatSettings.maxTokens}`);
 				text.inputEl.type = "number";
 				text.onChange((change) => {
-					this.plugin.settings.tokens = parseInt(change);
+					viewSettings.chatSettings.maxTokens = parseInt(change);
 					this.plugin.saveSettings();
 				});
 			});
+
+		if (modelType === "openAI") {
+			const frequencyPenalty = new Setting(parentContainer)
+				.setName("Frequency Penalty")
+				.setDesc(
+					"Defaults to 0. Number between -2.0 and 2.0. Positive values penalize new tokens based on their existing frequency in the text so far, decreasing the model's likelihood to repeat the same line verbatim."
+				)
+				.addText((text) => {
+					text.setValue(
+						`${viewSettings.chatSettings.openAI?.frequencyPenalty}`
+					);
+					text.inputEl.type = "number";
```

Comment:
Moving "number" to a const and referencing that const is a minor follow up improvement.

Reply:
(ideally the const lives in the types directory?) Or whatever is best practice in Typescript. I am new with TS.

Reply:
This i wasn't sure about this since its just an HTML input type like "email" or "password." Pretty much all my TS knowledge comes from googling on this project. 

We can definitely go that way if you think it makes more sense